File: 101-symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        def helper(root1, root2):
            if root1 is None and root2 is None:
                return True
            if root1 is None or root2 is None:
                return False
            if root1.val != root2.val:
                return False
            
            return helper(root1.left, root2.right) and helper(root1.right, root2.left)
        
        return helper(root, root)

File: 101-symmetric-tree/symmetric-tree.java
class Solution {

    boolean helper(TreeNode node1 , TreeNode node2) {
        if(node1 == null && node2 == null) {
            return true;
        } 
        if(node1 == null || node2 == null) {
            return false;
        }

        if(node1.val != node2.val) {
            return false;
        }


        boolean x = helper(node1.left , node2.right);
        boolean y = helper(node1.right , node2.left);

        if(x == true && y == true) {
            return true;
        }

        return false;
    }

    public boolean isSymmetric(TreeNode root) {
        return helper(root , root);
    }
}